lib/modules.js

```diff
@@ -126,59 +113,78 @@ async function getLastSeenModuleDistTag ({ module, moduleVersionsDir }) {
+export async function updateSourceFiles ({ module, ipnsKey, moduleVersionsDir }) {
   await mkdir(moduleBinaries, { recursive: true })
   const outDir = join(moduleBinaries, module)
 
-  const lastSeenDistTag = await getLastSeenModuleDistTag({ module, moduleVersionsDir })
-  if (lastSeenDistTag !== undefined) {
+  const lastSeenCID = await getLastSeenModuleCID({ module, moduleVersionsDir })
+  if (lastSeenCID !== undefined) {
     // Use `console.error` because with `--json` stdout needs to be JSON only
     console.error(`[${module}]  ⇣ checking for updates`)
   }
 
-  const distTag = await getLatestDistTag(repo)
-  const isUpdate = lastSeenDistTag !== distTag
+  const cid = await getLatestCID(ipnsKey)
+  const isUpdate = lastSeenCID !== cid
   if (!isUpdate) {
     console.error(`[${module}]  ✓ no update available`)
     return isUpdate
   }
 
-  console.error(`[${module}]  ⇣ downloading source files`)
-  const url = `https://github.com/${repo}/archive/refs/tags/${distTag}.tar.gz`
-  const res = await fetch(url, {
-    headers: {
-      ...(authorization ? { authorization } : {})
-    },
-    redirect: 'follow'
-  })
+  const url = `https://${cid}.ipfs.w3s.link?format=car`
+  console.error(`[${module}]  ⇣ downloading source files via ${url}`)
+  const res = await fetch(url)
 
   if (res.status >= 300) {
     throw new Error(
-      `[${module}] Cannot fetch ${module} archive for tag ${distTag}: ${res.status}\n` +
+      `[${module}] Cannot fetch ${module} archive for ${cid}: ${res.status}\n` +
       await res.text()
     )
   }
 
   if (!res.body) {
     throw new Error(
-      `[${module}] Cannot fetch ${module} archive for tag ${distTag}: no response body`
+      `[${module}] Cannot fetch ${module} archive for ${cid}: no response body`
     )
   }
 
   try {
-    // `{ strip: 1}` tells tar to remove the top-level directory (e.g. `mod-peer-checker-v1.0.0`)
-    await pipeline(res.body, gunzip(), tar.extract(outDir, { strip: 1 }))
+    const reader = await CarReader.fromIterable(res.body)
+    const entries = exporter(cid, {
+      async get (blockCid) {
+        const block = await reader.get(blockCid)
+        try {
+          await validateBlock(block)
+        } catch (err) {
+          throw new Error(
+            `Invalid block ${blockCid} of root ${cid}. ${err.stack}`
+          )
+        }
+        return block.bytes
+      }
+    })
+    const { value: entry } = await entries.next()
+    assert(entry, `No entries in ${module} archive`)
+    // Depending on size, entries might be packaged as `file` or `raw`
+    // https://github.com/web3-storage/w3up/blob/e8bffe2ee0d3a59a977d2c4b7efe425699424e19/packages/upload-client/src/unixfs.js#L11
+    if (entry.type === 'file' || entry.type === 'raw') {
+      // `{ strip: 1}` tells tar to remove the top-level directory (e.g. `mod-peer-checker-v1.0.0`)
+      await pipeline(
+        entry.content(),
+        gunzip(),
+        tar.extract(outDir, { strip: 1 })
```

Comment:
What happens when `ourDir` already exists and the new module version no longer contains some files? (E.g. we delete a source code file in Spark and then publish a new version.)

Are we going to end up with unused old source code files?

Reply:
good point! I think this can happen. https://github.com/filecoin-station/core/issues/336